scripts/render_notebook_viz.py

```
@@ -0,0 +1,290 @@
+    *,
+    verbose: bool,
+) -> None:
+    for i, cell in enumerate(nb.get("cells", [])):
+        if cell.get("cell_type") != "code":
+            continue
+        source = "".join(cell.get("source", []))
+        if not source.strip():
+            continue
+        try:
+            if verbose:
+                exec(compile(source, str(nb_path), "exec"), globals_ns)
```

Ruff (0.14.14)

[error] Use of `exec` detected

(S102)